src/types/module.rs

```diff
@@ -194,11 +195,50 @@ impl PyModule {
     /// ```rust,ignore
     /// m.add("also_double", wrap_pyfunction!(double)(py));
     /// ```
-    pub fn add_wrapped(&self, wrapper: &impl Fn(Python) -> PyObject) -> PyResult<()> {
-        let function = wrapper(self.py());
-        let name = function
-            .getattr(self.py(), "__name__")
-            .expect("A function or module must have a __name__");
-        self.add(name.extract(self.py()).unwrap(), function)
+    ///
+    /// **This function will be deprecated in the next release. Please use the specific
+    /// [add_function] and [add_submodule] functions instead.**
+    pub fn add_wrapped<'a, T>(&'a self, wrapper: &impl Fn(Python<'a>) -> T) -> PyResult<()>
+    where
+        T: IntoPyCallbackOutput<PyObject>,
+    {
+        let function = wrapper(self.py()).convert(self.py())?;
+        let name = function.getattr(self.py(), "__name__")?;
+        self.add(name.extract(self.py())?, function)
+    }
+
+    /// Add a submodule to a module.
+    ///
+    /// Use this together with `#[pymodule]` and [wrap_pymodule!].
+    ///
+    /// ```rust,ignore
+    /// m.add_submodule(wrap_pymodule!(utils));
+    /// ```
+    pub fn add_submodule<'a>(&'a self, wrapper: &impl Fn(Python<'a>) -> PyObject) -> PyResult<()> {
+        let module = wrapper(self.py());
+        let name = module.getattr(self.py(), "__name__")?;
+        self.add(name.extract(self.py())?, module)
+    }
+
+    /// Add a function to a module.
+    ///
+    /// Use this together with the`#[pyfunction]` and [wrap_pyfunction!].
+    ///
+    /// ```rust,ignore
+    /// m.add_function(wrap_pyfunction!(double));
+    /// ```
+    ///
+    /// You can also add a function with a custom name using [add](PyModule::add):
+    ///
+    /// ```rust,ignore
+    /// m.add("also_double", wrap_pyfunction!(double)(py, m));
```

Comment:
Does this need to handle the `Result` from wrap_pyfunction?

Reply:
Ah, good spot. This I think is an outdated doc and probably shows why using `ignore` in docstrings is usually a bad idea! Maybe this should be updated to not use ignore?

Reply:
Yeah, nice catch. Looks we need to update this.

Reply:
I made all the examples in the docs  of `fn add_X` runnable.